### ISSUE TITLE
Adding a NotFound Handler to a route group

### ISSUE THREAD

**Issue by tchapi** [slimphp/Slim#2220]
Hi,
I'm trying to add a specific NotFoundHandler (or NotAllowedHandler) to a specific group of routes. My use case is that I have an API sitting at `/api` and a frontend website at `/` and I want to be able to separate the not found handlers : one will be a JSON response with informative info, the other a nice HTML template with a "Ooops" message in it.

Is this possible ?
Thanks a lot

**Comment by l0gicgate** [comment]
You should have on notFoundHandler that handles both cases but provides a different response for each.

When a handler is being invoked, it gets passed Request, Response and Exception. You could have logic that looks at the Uri and decides whether it returns a JSON formatted response or calls your renderer to display some HTML depending on the path.

```php
class ErrorHandler
{
    public function __invoke(Request $request, Response $response, $exception)
    {
        $uri = $request->getUri();
        $path = $uri->getPath();
        $payload = [
            'code' => $exception->getCode(),
            'message' => $exception->getMessage();
        ];

        if (strstr($path, '/api') !== false) {
            return $response->withJson(
                $payload,
                $exception->getCode(),
                JSON_PRETTY_PRINT
            );
        } else {
            return $this->renderer->render($response, "/myError.php", $payload);
        }
    }
}

```

**Comment by tchapi** [comment]
Hi — thanks for your response. It definitely works but I can't use my middleware that does the same thing for normal pages, so it feels like duplicating the code for nothing. I was hoping that one could attach the handler per group or per route, a bit like a middleware.

Guess I'll create a kind of general service to stay DRY
thanks again
BR

**Comment by l0gicgate** [comment]
Would you be able to post a sample of your middleware so we can better understand what it does? I have a feeling that some code may not be in the right place hence why you'd have to duplicate it in the error handler.

**Comment by tchapi** [comment]
I'm using https://github.com/akrabat/rka-content-type-renderer to render different formats, and I would like my notFound or notAllowed handler to use them too, for the `/api` group, not for the other routes (not written here, but I have some other routes apart from the api routes).

Here is some code so you can understand the whole picture. The part I have a feeling I will need to duplicate is the middleware, that checks the format header, modify it according to params in case and do a few more things to the response afterwards

     $this->slimApp->group('/api', function () use ($renderer) {
            $this->get('/platforms', function($request, $response) {
                 $platforms = /* some array with data */
                 $response = $renderer->render($request, $response, $data);
                 return $response->withStatus(200);
           });
           /* more api routes & methods */
      })->add(function ($request, $response, $next) {
            $request = $request->withHeader('Accept', 'application/json'); // default
            $format = $request->getParam('format');
            if ($format) {
                $mapping = [
                    'html' => 'text/html',
                    'xml' => 'application/xml',
                    'json' => 'application/json',
                ];
                if (isset($mapping[$format])) {
                    $request = $request->withHeader('Accept', $mapping[$format]);
                }
            }

            $response = $next($request, $response);
          
           /* Some other things I do on the response */

            return $response;
      });

Thanks a lot for your time !

**Comment by l0gicgate** [comment]
I think that you are using the wrong approach here.
I would use global middleware instead of route specific middleware
If you could put a gist with all the moving parts together I would be able to assist you better also.

Here is an example:
```php
class ContentTypeMiddleware
{
    private $defaultContentType = 'application/json';
    private $allowedContentTypes = [
        'html' => 'text/html',
        'xml' => 'application/xml',
        'json' => 'application/json',
    ];

    public function __invoke(Request $request, Response $response, $next)
    {
        $request = $this->resolveContentType($request);

        /**
         * Process All Middleware
         */
        $response = $next($request, $response);

        /**
         * Set Accept Header Before Responding To The Request
         */
        $contentType = $request->getAttribute('contentType');
        $response = $response->withHeader('Accept', $contentType);

        return $response;
    }

    private function resolveContentType(Request $request)
    {
        $contentType = $this->defaultContentType;
        $resolvedContentType = $request->getParam('format');

        if (
            $resolvedContentType
            && isset($this->allowedContentTypes[$resolvedContentType])
        ) {
            $contentType = $resolvedContentType;
        }

        /**
         * Here we add an attribute to the request object 
         * in case we need to modify the final content type elsewhere in our application
         * The final Accept header will be set from this variable
         */
        $request->withAttribute('contentType', $contentType);

        return $request;
    }
}

/**
 * Add our middleware globally so every request will be processed with it
 * You could also use this middleware per route if you chose not to add it globally
 * eg.: $app->group(...)->add(new ContentTypeMiddleware);
 */
$app = new \Slim\App($settings);
$app->add(new ContentTypeMiddleware);

/**
 * Routes
 */
$app->group('/api', function () use ($renderer) {
    $this->get('/platforms', function ($request, $response) {
        $data = [];
        $response = $renderer->render($request, $response, $data);
        return $response->withStatus(200);
    });
});

/**
 * Run App
 */
$app->run();
```

**Comment by tchapi** [comment]
Thanks for the code, I understand the concept of the `$contentType` var that can be modified in the route to change the behavior, but my problem is within the **notFound** handlers, not the conventional routes.

In fact the underlying problem is that the **notFound** handlers do not have access to the route (it's empty, by design), so I cannot decide whether or not the **notFound** was triggered inside the api group or not. That's why I wanted a **notFound** handler _per_ group, but it doesn't exist;

Basically I'll do something like that in the middleware and in the **notFound** Handler I guess, though it's clearly not ideal : 

```php
$isApiRoute = (substr($request->getUri()->getPath(), 0, 4) === 'api/');
```

But it works. I have to duplicate the code, that's what I wanted to avoid but it seems unavoidable in my case.

**Comment by l0gicgate** [comment]
Did you set `determineRouteBeforeAppMiddleware` to `true` in your `App` settings?

That way you will have access to the route object in your middleware. 

Also, if you did a gist with the relevant parts of your code, it would be much easier for me to help. 

There's absolutetly no need to repeat code.

**Comment by geggleto** [comment]
This issue is really old. I believe it has also been answered; so I am going to close it. if you feel as if you still need support please reopen.
